packages/react/src/useSignUpSecurityKeyEmail.ts

```diff
@@ -0,0 +1,88 @@
+import {
+  SignUpSecurityKeyOptions,
+  signUpSecurityKeyPromise,
+  SignUpSecurityKeyState
+} from '@nhost/core'
+import { useSelector } from '@xstate/react'
+
+import { useAuthInterpreter } from './useAuthInterpreter'
+
+type SignUpSecurityKeyHandlerResult = Omit<SignUpSecurityKeyState, 'isLoading'>
+
+interface SignUpSecurityKeyHandler {
+  (email: string, options?: SignUpSecurityKeyOptions): Promise<SignUpSecurityKeyHandlerResult>
+}
+
+export interface SignUpSecurityKeyHookResult extends SignUpSecurityKeyState {
+  /** Used for a new user to sign up with a security key. Returns a promise with the current context */
+  signUpSecurityKeyEmail: SignUpSecurityKeyHandler
+}
+
+interface SignUpSecurityKeyHook {
+  (options?: SignUpSecurityKeyOptions): SignUpSecurityKeyHookResult
+}
+
+/**
+ * Use the hook `useSignUpSecurityKey` to sign up a user with security key and an email using the WebAuthn API.
+ *
+ * @example
+ * ```tsx
+ * const { signUpSecurityKey, needsEmailVerification, isLoading, isSuccess, isError, error } = useSignUpSecurityKey()
+ *
+ * console.log({ needsEmailVerification, isLoading, isSuccess, isError, error });
+ *
+ * const handleFormSubmit = async (e) => {
+ *   e.preventDefault();
+ *
+ *   await signUpSecurityKey('joe@example.com')
+ * }
+ * ```
+ *
+ * @docs https://docs.nhost.io/reference/react/use-sign-up-security-key
+ */
+export const useSignUpSecurityKeyEmail: SignUpSecurityKeyHook = (
+  hookOptions?: SignUpSecurityKeyOptions
+) => {
+  const service = useAuthInterpreter()
+  const isError = useSelector(service, (state) => !!state.context.errors.registration)
+
+  const error = useSelector(
+    service,
+    (state) => state.context.errors.registration || null,
+    (a, b) => a?.error === b?.error
+  )
+
+  const isLoading = useSelector(service, (state) => state.matches('registration.securityKey'))
+
+  const needsEmailVerification = useSelector(service, (state) =>
+    state.matches('registration.incomplete.needsEmailVerification')
+  )
+
+  const isSuccess = useSelector(service, (state) =>
+    state.matches({
+      authentication: 'signedIn',
+      registration: 'complete'
+    })
+  )
+
+  const signUpSecurityKeyEmail: SignUpSecurityKeyHandler = (email, options = hookOptions) =>
+    signUpSecurityKeyPromise(service, email, options)
+
+  const user = useSelector(
+    service,
+    (state) => state.context.user,
+    (a, b) => a?.id === b?.id
+  )
+  const accessToken = useSelector(service, (state) => state.context.accessToken.value)
+
+  return {
+    accessToken,
+    error,
+    isError,
```

Comment:
I was thinking about this for a while now. What do you think about introducing a `status` value besides `isLoading` / `isSuccess` and `isError` at some point in the future? It could hold values any of these values `loading | success | `error`. (See [TanStack Query's code](https://github.com/TanStack/query/blob/900d2c51306a91372ae38d0d6846ae61a165c326/packages/query-core/src/query.ts#L488) where something similar is implemented).

It would make it more convenient to use these hooks for React users in some cases.

For example instead of this:

```ts
function Component() {
  const { isLoading, isSuccess, isError, signUp } = useSignUp()
  
  useEffect(() => {
    if (isLoading) { ... }
    if (isError) { ... }
    if (isSuccess) { ... }
  }, [isLoading, isSuccess, isError, /* Many other things in this dep. array */])

  render <SomeComponent />
}
```

Users could use something like this:
```tsx
function Component() {
  const { status, signUp } = useSignUp()
  
  useEffect(() => {
    if (status === 'loading') { ... }
    if (status === 'error') { ... }
    if (status === 'success') { ... }
  }, [status, /* Many other things in this dep. array */])

  render <SomeComponent />
}
```
Maintaining huge dependency arrays is a mental overhead which we can help by eliminating the need to use three separate values. [This blog](https://kentcdodds.com/blog/stop-using-isloading-booleans) post also talks about this topic. I find it very interesting.

Reply:
I like the idea and also feel our way of doing things is not ideal

Reply:
@szilarddoro Feel free to create a GitHub issue for this. 👍